### PR TITLE
ci: Add HEAD of dependencies workflow

### LINKED ISSUE #373: Add HEAD of dependencies GHA workflow
Hi @andrzejnovak. Now that `matplotlib` has public nightly wheels (c.f. https://github.com/matplotlib/matplotlib/pull/22733) would you be interested in a "HEAD of dependencies workflow" that runs nightly to test against the `HEAD` on GitHub or nightlies of the core dependencies

https://github.com/scikit-hep/mplhep/blob/7a20bbd56b5ca4e9029a26400ae0873afc77badd/setup.cfg#L27-L32

? If so, I can basically port over [the one we have for `pyhf`](https://github.com/scikit-hep/pyhf/blob/0a8f3282e9bc6d66b3e596fd35c74a6e03ff567a/.github/workflows/dependencies-head.yml) and PR it for you here.

### PR DESCRIPTION
**Pull request by matthewfeickert**
Resolves #373 

Test against changes to core dependencies using pre-releases on PyPI, the HEAD of dependencies on GitHub, or nightly release wheels.

If any of the tests with these updated dependencies fail, upload the pytest test results as artifacts for inspection.

### CODE REVIEW

File: .github/workflows/head-dependencies.yml [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        python -m pip install --upgrade pip setuptools wheel
+        python -m pip --no-cache-dir --quiet install --upgrade .[test]
+        python -m pip install --pre --upgrade --extra-index-url https://pypi.anaconda.org/scipy-wheels-nightly/simple matplotlib
+
+    - name: List installed Python packages
+      run: python -m pip list
+
+    - name: Test with pytest
+      run: |
+        pytest -r sa --mpl --mpl-results-path=pytest_results
```
Comment (matthewfeickert) [inline .github/workflows/head-dependencies.yml:76]: When I checked this with a test run `test_histplot_kwargs` is failing against the nightly build wheel. Something to keep on the radar https://github.com/scikit-hep/mplhep/runs/5790229209?check_suite_focus=true

Comment (matthewfeickert) [reply .github/workflows/head-dependencies.yml:86]: Done! :)

Comment (matthewfeickert) [reply .github/workflows/head-dependencies.yml:86]: Oh wait. When I ran on a debug branch I got

```
Warning: No files were found with the provided path: pytest_results. No artifacts will be uploaded.
```

let me look into that.

**Comment by andrzejnovak** [comment]
Thanks a lot @matthewfeickert!

File: .github/workflows/head-dependencies.yml (hunk 1 above)
Comment (matthewfeickert) [reply .github/workflows/head-dependencies.yml:86]: Issue #375.

**Pull request merged**